File: infrastructures/ddd/csharp/src/Infrastructure/Queues/Impl/InMemoryEventBus.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace OddlyDdd.Infrastructure.Queues.Impl
{
    /// <summary>
    /// In-memory implementation of event bus for demonstration purposes.
    /// In production, replace with actual message queue (RabbitMQ, Azure Service Bus, AWS SQS, etc.)
    /// 
    /// This implementation:
    /// - Lives in /Infrastructure/Queues/Impl/
    /// - Implements IEventPublisher and IEventSubscriber
    /// - Provides simple pub/sub mechanism
    /// - NOT suitable for production (no persistence, no retry, no scaling)
    /// </summary>
    public class InMemoryEventBus : IEventPublisher, IEventSubscriber
    {
        private readonly ConcurrentDictionary<string, List<Func<object, Task>>> _subscribers = new();

        /// <summary>
        /// Publishes an event to all subscribers of the topic.
        /// Topic naming: {subdomain}.{action} (e.g., "example.created")
        /// </summary>
        public async Task PublishAsync<TEvent>(TEvent p_event, string p_topic)
        {
            if (!_subscribers.TryGetValue(p_topic, out var handlers))
            {
                return; // No subscribers for this topic
            }

            var tasks = new List<Task>();
            foreach (var handler in handlers)
            {
                tasks.Add(handler(p_event!));
            }

            await Task.WhenAll(tasks);
        }

        /// <summary>
        /// Subscribes to events on the specified topic.
        /// Topic naming: {subdomain}.{action} (e.g., "example.created")
        /// </summary>
        public Task SubscribeAsync<TEvent>(string p_topic, Func<TEvent, Task> p_handler)
        {
            _subscribers.AddOrUpdate(
                p_topic,
                new List<Func<object, Task>> { obj => p_handler((TEvent)obj) },
                (key, existing) =>
                {
                    existing.Add(obj => p_handler((TEvent)obj));
                    return existing;
                }
            );

            return Task.CompletedTask;
        }
    }
}

File: infrastructures/ddd/java/src/main/java/com/oddly/ddd/infrastructure/queues/InMemoryEventBus.java
package com.oddly.ddd.infrastructure.queues;

import com.oddly.ddd.infrastructure.queues.infra.IEventPublisher;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import java.util.concurrent.CompletableFuture;

/**
 * In-memory implementation of event bus for demonstration.
 * Queue implementations:
 * - Live in /infrastructure/queues/
 * - Implementations in root (NO /impl/ - no autowiring contracts)
 * - For production, replace with RabbitMQ, Kafka, etc.
 */
@Component
public class InMemoryEventBus implements IEventPublisher {
    private static final Logger LOGGER = LoggerFactory.getLogger(InMemoryEventBus.class);

    @Override
    public <TEvent> CompletableFuture<Void> publishAsync(TEvent p_event, String p_topic) {
        return CompletableFuture.runAsync(() -> {
            // In a real implementation, this would publish to a message queue
            LOGGER.info("Publishing event to topic '{}': {}", p_topic, p_event.getClass().getSimpleName());
            // For demonstration, just log the event
        });
    }
}
